tests/net/socket/tls/src/main.c

```diff
@@ -0,0 +1,244 @@
+
+	int r;
+	int server_fd;
+	int client_fd;
+	char *addrstrp;
+	k_tid_t server_thread_id;
+	struct sockaddr_in sa;
+	char addrstr[INET_ADDRSTRLEN];
+
+	k_thread_name_set(k_current_get(), "ztest_thread");
+
+	/* Load credentials */
+	NET_DBG("Loading credentials");
+	r = tls_credential_add(SERVER_CERTIFICATE_TAG,
+		TLS_CREDENTIAL_CA_CERTIFICATE,
+		ca, sizeof(ca));
+	zassert_equal(r, 0, "failed to add CA Certificate (%d)", r);
+
+	r = tls_credential_add(SERVER_CERTIFICATE_TAG,
+		TLS_CREDENTIAL_SERVER_CERTIFICATE,
+		server, sizeof(server));
+	zassert_equal(r, 0, "failed to add Server Certificate (%d)", r);
+
+	r = tls_credential_add(SERVER_CERTIFICATE_TAG,
+		TLS_CREDENTIAL_PRIVATE_KEY,
+		server_privkey, sizeof(server_privkey));
+	zassert_equal(r, 0, "failed to add Server Private Key (%d)", r);
+
+	/* Create server socket */
+	NET_DBG("Creating server socket");
+	r = socket(AF_INET, SOCK_STREAM, IPPROTO_TLS_1_2);
+	zassert_not_equal(r, -1, "failed to create server socket (%d)", errno);
+	server_fd = r;
+
+    r = setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes));
+    zassert_not_equal(r, -1, "failed to set SO_REUSEADDR (%d)", errno);
+
+	r = setsockopt(server_fd, SOL_TLS, TLS_SEC_TAG_LIST,
+		sec_tag_list, sizeof(sec_tag_list));
+	zassert_not_equal(r, -1, "failed to set TLS_SEC_TAG_LIST (%d)", errno);
+
+	memset(&sa, 0, sizeof(sa));
+	r = inet_pton(AF_INET, CONFIG_NET_CONFIG_MY_IPV4_ADDR,
+		&sa.sin_addr.s_addr);
+	sa.sin_family = AF_INET;
+	sa.sin_port = htons(PORT);
+
+	r = bind(server_fd, (struct sockaddr *)&sa, sizeof(sa));
+	zassert_not_equal(r, -1, "failed to bind (%d)", errno);
+
+	r = listen(server_fd, 1);
+	zassert_not_equal(r, -1, "failed to listen (%d)", errno);
+
+    memset(addrstr, '\0', sizeof(addrstr));
+    addrstrp = (char *)inet_ntop(AF_INET, &sa.sin_addr,
+        addrstr, sizeof(addrstr));
+    zassert_not_equal(addrstrp, NULL, "inet_ntop() failed (%d)", errno);
+
+    NET_DBG("listening on [%s]:%d as fd %d",
+        log_strdup(addrstr), ntohs(sa.sin_port), server_fd);
+
+	/* Launch the server thread */
+    NET_DBG("Launching server thread");
+	server_thread_id = k_thread_create(&server_thread, server_stack, STACK_SIZE,
+		server_thread_fn, (void *)(intptr_t)server_fd, NULL, NULL, K_PRIO_PREEMPT(8),
+		0, K_NO_WAIT);
+	zassert_equal(r, 0, "failed to start thread (%d)", r);
+
+	k_thread_name_set(server_thread_id, "server");
+	k_thread_name_set(k_current_get(), "client");
+
+	/* Create client socket */
+	NET_DBG("Creating client socket");
+	r = socket(AF_INET, SOCK_STREAM, IPPROTO_TLS_1_2);
+	zassert_not_equal(r, -1, "failed to create client socket (%d)", errno);
+	client_fd = r;
+
+	r = setsockopt(client_fd, SOL_TLS, TLS_SEC_TAG_LIST,
+		sec_tag_list, sizeof(sec_tag_list));
+	zassert_not_equal(r, -1, "failed to set TLS_SEC_TAG_LIST (%d)", errno);
+
+	NET_DBG("Calling connect()");
+	r = connect(client_fd, (struct sockaddr *)&sa, sizeof(sa));
+	zassert_not_equal(r, -1, "failed to connect (%d)", errno);
+
```

Comment:
After looking how `tests/net/socket/tcp` is doing things when connecting, I think you need to let other threads to run too, so `k_msleep()` might help here.

Reply:
Good call. I was occasionally unable to connect, although @rlubos hinted that may have been due to not having enough buffers / packets.

Reply:
Definitely seeing things hanging on `mps2_an385`, but adding a `k_msleep()` after connect does not seem to help.